File: 198.house_robber/rob.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var rob = function(nums) {
    if(nums.length === 0) return 0;
    var maxValue = [];
    maxValue[0] = nums[0];
    maxValue[1] = Math.max(nums[1], nums[0]);
    for(var i = 2; i < nums.length; i++){
        maxValue[i] = Math.max(maxValue[i-1], maxValue[i-2] + nums[i]);//take this value or not
    }
    return maxValue[nums.length-1];
};

console.log(rob([1,4,2,8]));

File: 198.house_robber/rob.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        prev_take, prev_nottake = 0, 0
        for i in nums:
            prev_take, prev_nottake = prev_nottake + \
                i, max(prev_take, prev_nottake)
        return max(prev_take, prev_nottake)


'''
record the maximum gain based on whether robbing the current house or not
'''
